File: packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/base/AtomicResult.kt
/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
package com.amazonaws.amplify.amplify_auth_cognito.base

import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.*

/**
 * Thread-safe [MethodChannel.Result] wrapper which prevents multiple replies and automatically posts
 * results to the main thread.
 */
class AtomicResult(private val result: MethodChannel.Result) : MethodChannel.Result {
    private companion object {
        /**
         * Scope for performing result handling.
         * Method channel results must be sent on the main (UI) thread.
         */
        val scope = MainScope()
    }

    /**
     * Whether a response has been sent.
     */
    private var isSent = false

    override fun success(value: Any?) {
        scope.launch {
            if (isSent) return@launch
            result.success(value)
            isSent = true
        }
    }

    override fun error(errorCode: String?, errorMessage: String?, errorDetails: Any?) {
        scope.launch {
            if (isSent) return@launch
            result.error(errorCode, errorMessage, errorDetails)
            isSent = true
        }
    }

    override fun notImplemented() {
        scope.launch {
            if (isSent) return@launch
            result.notImplemented()
            isSent = true
        }
    }
}

File: packages/amplify_core/ios/Classes/Support/AtomicResult.swift
/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import Foundation

// swiftlint:disable identifier_name type_name

/// Thread-safe wrapper for [FlutterResult]. Prevents multiple replies and automically posts results to the main thread.
public func AtomicResult(_ result: @escaping FlutterResult, _ methodName: String) -> FlutterResult {
    return atomicResult(result, methodName).send
}

private class atomicResult {
    let result: FlutterResult

    /// The method call which initiated this result.
    let methodName: String

    /// Whether a reply has already been sent.
    var isSent = false

    init(_ result: @escaping FlutterResult, _ methodName: String) {
        self.result = result
        self.methodName = methodName
    }

    func send(_ value: Any?) {
        DispatchQueue.main.async { [self] in
            guard !isSent else {
                NSLog("AtomicResult(%@): Attempted to send value after initial reply", methodName)
                return
            }
            result(value)
            isSent = true
        }
    }
}
